# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/ArrayBufferObject.cpp
# Commit: 40c6d6eed7f8
# Full Hash: 40c6d6eed7f8ffd79078b778b6ceb8e78b28c2f0
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2023-09-12 04:12:49
# Regressor Bug: 1690111
# File Overlap Count: 2
# Description:
#   Bug 1690111 - Implement JS::EnsureNonInlineArrayBufferOrView. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D176596
# ==============================================================================

diff -r 692f2a759573 -r 40c6d6eed7f8 js/src/vm/ArrayBufferObject.cpp
--- a/js/src/vm/ArrayBufferObject.cpp	Mon Sep 11 12:52:19 2023 +0000
+++ b/js/src/vm/ArrayBufferObject.cpp	Mon Sep 11 12:52:19 2023 +0000
@@ -1950,6 +1950,55 @@
 }
 
 /* static */
+bool ArrayBufferObject::ensureNonInline(JSContext* cx,
+                                        Handle<ArrayBufferObject*> buffer) {
+  if (buffer->isDetached()) {
+    return true;
+  }
+
+  if (buffer->isLengthPinned()) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_ARRAYBUFFER_LENGTH_PINNED);
+    return false;
+  }
+
+  MOZ_ASSERT(!buffer->isPreparedForAsmJS());
+
+  BufferContents inlineContents = buffer->contents();
+  if (inlineContents.kind() != INLINE_DATA) {
+    return true;
+  }
+
+  size_t nbytes = buffer->byteLength();
+  ArrayBufferContents copy = NewCopiedBufferContents(cx, buffer);
+  if (!copy) {
+    return false;
+  }
+  BufferContents outOfLineContents =
+      BufferContents::createMalloced(copy.release());
+  buffer->setDataPointer(outOfLineContents);
+  AddCellMemory(buffer, nbytes, MemoryUse::ArrayBufferContents);
+
+  if (!buffer->firstView()) {
+    return true;  // No views! Easy!
+  }
+
+  buffer->firstView()->as<ArrayBufferViewObject>().notifyBufferMoved(
+      inlineContents.data(), outOfLineContents.data());
+
+  auto& innerViews = ObjectRealm::get(buffer).innerViews.get();
+  if (InnerViewTable::ViewVector* views =
+          innerViews.maybeViewsUnbarriered(buffer)) {
+    for (JSObject* view : *views) {
+      view->as<ArrayBufferViewObject>().notifyBufferMoved(
+          inlineContents.data(), outOfLineContents.data());
+    }
+  }
+
+  return true;
+}
+
+/* static */
 void ArrayBufferObject::addSizeOfExcludingThis(
     JSObject* obj, mozilla::MallocSizeOf mallocSizeOf, JS::ClassInfo* info,
     JS::RuntimeSizes* runtimeSizes) {